sql-plugin/src/main/scala/org/apache/spark/sql/rapids/collectionOperations.scala

```diff
@@ -407,260 +411,237 @@ class GpuSequenceMeta(
+   * a null in the output.
+   *
+   * The returned column should be closed.
    */
-  private def calculateSizeAndStep(start: BinaryOperable, stop: BinaryOperable, dt: DataType):
-      Seq[ColumnVector] = {
-    withResource(stop.sub(start)) { difference =>
-      withResource(GpuSequenceUtil.numberScalar(dt, 1)) { one =>
-        val step = withResource(GpuSequenceUtil.numberScalar(dt, -1)) { negativeOne =>
-          withResource(GpuSequenceUtil.numberScalar(dt, 0)) { zero =>
-            withResource(difference.greaterOrEqualTo(zero)) { pred =>
-              pred.ifElse(one, negativeOne)
-            }
-          }
+  def computeSequenceSizes(
+      start: ColumnVector,
+      stop: ColumnVector,
+      step: ColumnVector): ColumnVector = {
+    // Keep the same requirement with Spark:
+    // (step > 0 && start <= stop) || (step < 0 && start >= stop) || (step == 0 && start == stop)
+    withResource(Scalar.fromByte(0.toByte)) { zero =>
+      // The check should ignore each row (Row(start, stop, step)) that contains at least
+      // one null element according to Spark's code. Thanks to the cudf binary ops, who ignore
+      // nulls already, skipping nulls can be done without any additional process.
+      //
+      // Because the filtered table (e.g. upTbl) in each rule check excludes the rows that the
+      // step is null. Next a null row will be produced when comparing start or stop when any
+      // of them is null, and the nulls are skipped in the final assertion 'isAllValidTrue'.
+      withResource(new Table(start, stop)) { startStopTable =>
+        // (step > 0 && start <= stop)
+        val upTbl = withResource(step.greaterThan(zero)) { positiveStep =>
+          startStopTable.filter(positiveStep)
+        }
+        val allUp = withResource(upTbl) { _ =>
+          upTbl.getColumn(0).lessOrEqualTo(upTbl.getColumn(1))
+        }
+        withResource(allUp) { _ =>
+          require(isAllValidTrue(allUp), "Illegal sequence boundaries: step > 0 but start > stop")
+        }
+
+        // (step < 0 && start >= stop)
+        val downTbl = withResource(step.lessThan(zero)) { negativeStep =>
+          startStopTable.filter(negativeStep)
+        }
+        val allDown = withResource(downTbl) { _ =>
+          downTbl.getColumn(0).greaterOrEqualTo(downTbl.getColumn(1))
+        }
+        withResource(allDown) { _ =>
+          require(isAllValidTrue(allDown),
+              "Illegal sequence boundaries: step < 0 but start < stop")
+        }
+
+        // (step == 0 && start == stop)
+        val equalTbl = withResource(step.equalTo(zero)) { zeroStep =>
+          startStopTable.filter(zeroStep)
+        }
+        val allEq = withResource(equalTbl) { _ =>
+          equalTbl.getColumn(0).equalTo(equalTbl.getColumn(1))
+        }
+        withResource(allEq) { _ =>
+          require(isAllValidTrue(allEq),
+              "Illegal sequence boundaries: step == 0 but start != stop")
+        }
+      }
+    } // end of zero
```

Comment:
This method is pretty big, and this is just pre-checks to the interesting part of the algorithm.  Recommend factoring this out into a private method to make the main method easier to follow.

Reply:
Good suggestion, updated.